doc/guide/transactions-and-threading.rst

```diff
@@ -0,0 +1,394 @@
+   will provide greater control over conflict resolution, including
+   the option of disabling it.
+
+   It's generally best to avoid conflicts in the first place, if possible.
+
+ZODB and atomicity
+==================
+
+ZODB provides atomic transactions. When using ZODB, it's important to
+align work with transactions.  Once a transaction is committed, it
+can't be rolled back [#undo]_ automatically.  For applications, this
+implies that work that should be atomic shouldn't be split over
+multiple transactions.  This may seem somewhat obvious, but the rule
+can be broken in non-obvious ways. For example a Web API that splits
+logical operations over multiple web requests, as is often done in
+`REST
+<https://en.wikipedia.org/wiki/Representational_state_transfer>`_
+APIs, violate this rule.
+
+Partial transaction error recovery using savepoints
+---------------------------------------------------
+
+A transaction can be split into multiple steps that can be rolled back
+individually.  This is done by creating savepoints.  Changes in a
+savepoint can be rolled back without rolling back an entire
+transaction::
+
+  import ZODB
+  db = ZODB.DB(None) # using a mapping storage
+  with db.transaction() as conn:
+      conn.root.x = 1
+      conn.root.y = 0
+      savepoint = conn.transaction_manager.savepoint()
+      conn.root.y = 2
+      savepoint.rollback()
+
+  with db.transaction() as conn:
+      print([conn.root.x, conn.root.y]) # prints 1 0
+
+.. -> src
+
+   >>> exec(src)
+   [1, 0]
+
+If we executed this code, it would print 1 and 0, because while the
+initial changes were committed, the changes in the savepoint were
+rolled back.
+
+A secondary benefit of savepoints is that they save any changes made
+before the savepoint to a file, so that memory of changed objects can
+be freed if they aren't used later in the transaction.
+
+Concurrency, threads and processes
+==================================
+
+ZODB supports concurrency through transactions.  Multiple programs can
+operate independently in separate transactions.  They synchronize at
+transaction boundaries.
+
+The most common way to run ZODB is with each program running in it's
+own thread.  Usually using the thread-local transaction manager is used.
+
+You can use multiple threads per transaction and you can run multiple
+transactions in a single thread. To do this, you need to instantiate
+and use your own transaction manager, as described in `Explicit
+transaction managers`_.  To run multiple transaction managers
+simultaneously in a thread, you need to use a separate transaction
+manager for each transaction.
+
+To spread a transaction over multiple threads, you need to keep in
+mind that database connections, transaction managers and transactions
+are **not thread-safe**.  You have to prevent simultaneous access from
+multiple threads.  For this reason, **using multiple threads with a
+single connection is not recommended**, but it is possible with care.
```

Comment:
Hard to over-emphasize that point. It might even deserve a call-ahead in the "Explicit transaction manager" section.